nemo/collections/nlp/models/language_modeling/megatron/bert/bert_model.py

```diff
@@ -0,0 +1,745 @@
+    """Masked LM head for Bert
+
+    Arguments:
+        mpu_vocab_size: model parallel size of vocabulary.
+        hidden_size: hidden size
+        init_method: init method for weight initialization
+        layernorm_epsilon: tolerance for layer norm divisions
+        parallel_output: whether output logits being distributed or not.
+    """
+
+    def __init__(
+        self,
+        config: ModelParallelConfig,
+        mpu_vocab_size,
+        hidden_size,
+        init_method,
+        layernorm_epsilon,
+        parallel_output,
+        use_openai_gelu,
+        onnx_safe,
+    ):
+
+        super(BertLMHead, self).__init__(config=config)
+
+        self.bias = torch.nn.Parameter(torch.zeros(mpu_vocab_size))
+        set_tensor_model_parallel_attributes(self.bias, True, 0, 1)
+        self.parallel_output = parallel_output
+        self.sequence_parallel = config.sequence_parallel
+
+        self.dense = get_linear_layer(hidden_size, hidden_size, init_method)
+        self.layernorm = get_layer_norm(hidden_size, eps=layernorm_epsilon)
+        self.gelu = torch.nn.functional.gelu
+        if use_openai_gelu:
+            self.gelu = openai_gelu
+        elif onnx_safe:
+            self.gelu = erf_gelu
+
+    def forward(self, hidden_states, word_embeddings_weight):
+        hidden_states = self.dense(hidden_states)
+        hidden_states = self.gelu(hidden_states)
+        hidden_states = self.layernorm(hidden_states)
+        async_tensor_model_parallel_allreduce = self.config.async_tensor_model_parallel_allreduce
+        output = parallel_lm_logits(
+            hidden_states,
+            word_embeddings_weight,
+            self.parallel_output,
+            sequence_parallel=self.sequence_parallel,
+            bias=self.bias,
+            async_tensor_model_parallel_allreduce=async_tensor_model_parallel_allreduce,
+        )
+        return output
+
+
+def post_language_model_processing(
+    lm_output, pooled_output, lm_head, binary_head, lm_labels, logit_weights, fp16_lm_cross_entropy,
+):
+    # lm_logits: [s, b, vocab_size]
+    lm_logits = lm_head(lm_output, logit_weights)
+
+    binary_logits = None
+    if binary_head is not None:
+        # binary_logits: [s, b, 2] or [s, b, vocab_size] if binary_head is Identity
+        binary_logits = binary_head(pooled_output)
+
+    if lm_labels is None:
+        return lm_logits, binary_logits
+    else:
+        # match shape of labels to lm_logits
+        # lm_labels: [b, s] -> [s, b]
+        lm_labels = lm_labels.transpose(0, 1).contiguous()
+        if fp16_lm_cross_entropy:
+            assert lm_logits.dtype == torch.half
+            lm_loss = tensor_parallel.vocab_parallel_cross_entropy(lm_logits, lm_labels)
+        else:
+            lm_loss = tensor_parallel.vocab_parallel_cross_entropy(lm_logits.float(), lm_labels)
+        # lm_loss: [s, b]
+        return lm_loss, binary_logits
+
+
+@dataclass
+class TransformerLayerSubmodulesPostLNSupport:
```

Comment:
Can you just inherit from the transformerlayer class in mcore, and then add the extra stuff alone ?

Reply:
done